lib/AST/Decl.cpp

```diff
@@ -1433,6 +1433,35 @@ AbstractStorageDecl::getAccessStrategy(AccessSemantics semantics,
   llvm_unreachable("bad access semantics");
 }
 
+static bool hasPrivateOrFilePrivateFormalAccess(const ValueDecl *D) {
+  return D->hasAccess() && D->getFormalAccess() <= AccessLevel::FilePrivate;
+}
+
+/// Returns true if one of the ancestor DeclContexts of this ValueDecl is either
+/// marked private or fileprivate or is a local context.
+static bool isInPrivateOrLocalContext(const ValueDecl *D) {
+  const DeclContext *DC = D->getDeclContext();
+  if (!DC->isTypeContext()) {
+    assert((DC->isModuleScopeContext() || DC->isLocalContext()) &&
+           "unexpected context kind");
+    return DC->isLocalContext();
+  }
+
+  auto *nominal = DC->getAsNominalTypeOrNominalTypeExtensionContext();
+  if (nominal == nullptr)
+    return false;
+
+  bool inPrivateOrLocalContext = isInPrivateOrLocalContext(nominal);
+  if (hasPrivateOrFilePrivateFormalAccess(nominal) && !inPrivateOrLocalContext)
```

Comment:
You don't actually need to check `!inPrivateOrLocalContext`; you're going to return true on the next line anyway. And then it's cheaper to delay the recursive call.

Reply:
right.